src/server/pfs/server/driver_new_storage.go

```diff
@@ -0,0 +1,219 @@
+package server
+
+import (
+	"bytes"
+	"io"
+	"log"
+	"path"
+	"strconv"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/auth"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	pfsserver "github.com/pachyderm/pachyderm/src/server/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/chunk"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/tar"
+	txnenv "github.com/pachyderm/pachyderm/src/server/pkg/transactionenv"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"golang.org/x/net/context"
+)
+
+const (
+	compacted = "compacted"
+)
+
+func init() {
+	// This is kind of weird, but I think it might be necessary.
+	// Since we build our protos with gogoproto, we need to register
+	// the types that can be embedded in an Any type with the gogoproto
+	// proto package because building the protos just registers them with the
+	// normal proto package.
+	proto.RegisterType((*pfs.Merge)(nil), "pfs.Merge")
+	proto.RegisterType((*pfs.Shard)(nil), "pfs.Shard")
+}
+
+var (
+	// (bryce) need to expose as configuration.
+	memThreshold int64 = 1024 * chunk.MB
+)
+
+func (d *driver) startCommitNewStorageLayer(txnCtx *txnenv.TransactionContext, id string, parent *pfs.Commit, branch string, provenance []*pfs.CommitProvenance, description string) (*pfs.Commit, error) {
+	commit, err := d.startCommit(txnCtx, id, parent, branch, provenance, description)
+	if err != nil {
+		return nil, err
+	}
+	d.fs = d.storage.New(context.Background(), commit.ID, fileset.WithMemThreshold(memThreshold))
```

Comment:
`txnCtx` has a `Context` in it that should presumably be used here.

Reply:
I guess. I wanted to confirm that this was not another situation where the context isn't cancelled when the request is cancelled. The whole having to get a new pach client with a ctx rather than just propagating the ctx is error prone, and we have seen this issue with the job logic where the pach client ctx was not correct. I don't really understand the transaction stuff yet, so that added another layer to this concern.

Reply:
I'm not totally sure either, I think the ctx here will get cancelled with the request, @Tryneus would be the one that knows for sure. I think in job logic the issue was that there are a lot of places where it's actually an orphaned goro doing stuff rather than the main request goro.

Reply:
I need to make a few more changes to have this work with the correct context. I think those would fit better in a separate PR for now. This PR is more focused on the distributed merge implementation with the minimum necessary integration to test it.